tinygrad/shape/shapetracker.py

```diff
@@ -87,40 +99,79 @@ def view_from_shape(shape:Tuple[int, ...]) -> View:
   if vm2.mask: return None  # this isn't supported yet
   new_strides, new_offset = [], vm2.expr_node(Variable.num(vm1.offset))
-  assert isinstance(new_offset, NumNode), "new_offset wasn't a number?!?"
+  assert new_offset.__class__ == NumNode, "new_offset wasn't a number?!?"
   for s,st in zip(vm1.shape, vm1.strides):
     this_dim = View(vm2.shape, vm2.strides).expr_node(Variable('idx', 0, s-1)*st)
     if s == 1:
       new_strides.append(0)   # all shape 1 can have stride 0
-    elif isinstance(this_dim, NumNode) and this_dim.b == 0:
+    elif this_dim.__class__ == NumNode and this_dim.b == 0:
       new_strides.append(0)
-    elif isinstance(this_dim, Variable):
+    elif this_dim.__class__ == Variable:
       new_strides.append(1)
-    elif isinstance(this_dim, MulNode) and isinstance(this_dim.a, Variable):
+    elif this_dim.__class__ == MulNode and cast(MulNode, this_dim).a.__class__ == Variable:
       new_strides.append(this_dim.b)
     else:
       if DEBUG >= 4: print("can't simplify", s, this_dim.render())
       break
   return View(vm1.shape, tuple(new_strides), new_offset.b, vm1.mask) if len(new_strides) == len(vm1.strides) else None
 
+@functools.lru_cache(maxsize=None)
+def _reshape(view: View, new_shape: Tuple[int, ...]) -> Tuple[View, bool]:
+  shape, mask, strides, offset = view.shape, view.mask, view.strides, view.offset
+  # check if this is adding or removing 1s (only)
+  # NOTE: this is optional, but removes most calls to (expensive!) merge_views (with mask, not optional)
+  if [x for x in shape if x != 1] == [x for x in new_shape if x != 1]:
+    new_strides: List[int] = [y for x,y in zip(shape, strides) if x != 1]
+    new_strides_tuple: Tuple[int, ...] = tuple([0 if x == 1 else new_strides.pop(0) for x in new_shape])
+    new_mask_tuple = None
+    if mask:
+      for x,y in zip(shape, mask):
+        if x == 1 and y != (0, 1):
+          new_mask_tuple = tuple([(0,0) for _ in new_shape])
+          break
+      else:
+        new_mask: List[Tuple[int, int]] = [y for x,y in zip(shape, mask) if x != 1]
+        new_mask_tuple = tuple([(0,1) if x == 1 else new_mask.pop(0) for x in new_shape])
+    return View(new_shape, new_strides_tuple, offset, new_mask_tuple), False
+
+  new_view = View(new_shape, strides_for_shape(new_shape))
+  if view.contiguous: return new_view, False # NOTE: if it's contiguous it can't have an offset
+  else:
+    if (merged_view := merge_views(view, new_view)) is not None: return merged_view, False
+    else:
+      if DEBUG >= 4: print(f"WARNING: creating new view with reshape {view} -> {new_shape}")
+      return new_view, True
+
+@functools.lru_cache(maxsize=None)
+def get_pad_args(shape, arg: Tuple[Tuple[int, int], ...]):
+  return tuple([(-b,s+e) for s,(b,e) in zip(shape, arg)]), tuple([(b,s+b) for s,(b,_) in zip(shape, arg)])
+
+@functools.lru_cache(maxsize=None)
+def get_unsafe_resize_offset(strides, arg):
+  return sum(map(operator.mul, strides, map(operator.itemgetter(0), arg)))
```

Comment:
`sum([s*x[0] for s,x in zip(strides,arg)])` is a lot more readable, is it slower?

Reply:
Hmm yeah I can revert that, the story behind this is that I changed it before i decided to cache the whole thing and the gain was non-trivial because it was called so often, now the gain is minimal with caching on top